src/Microsoft.DotNet.Cli.Utils/Telemetry.cs

```diff
@@ -0,0 +1,117 @@
+
+            if (optout)
+            {
+                return;
+            }
+
+            try
+            {
+                _client = new TelemetryClient();
+                _client.InstrumentationKey = InstrumentationKey;
+                _client.Context.Session.Id = Guid.NewGuid().ToString();
+
+                var runtimeEnvironment = PlatformServices.Default.Runtime;
+                _client.Context.Device.OperatingSystem = runtimeEnvironment.OperatingSystem;
+
+                _commonProperties = new Dictionary<string, string>();
+                _commonProperties.Add(OSVersion, runtimeEnvironment.OperatingSystemVersion);
+                _commonProperties.Add(OSPlatform, runtimeEnvironment.OperatingSystemPlatform.ToString());
+                _commonProperties.Add(RuntimeId, runtimeEnvironment.GetRuntimeIdentifier());
+                _commonProperties.Add(ProductVersion, Product.Version);
+                _commonMeasurements = new Dictionary<string, double>();
+
+                _isInitialized = true;
+            }
+            catch (Exception)
+            {
+                // we dont want to fail the tool if telemetry fais. We should be able to detect abnormalities from data 
+                // at the server end
+            }
+        }
+
+        public void TrackEvent(string eventName, IDictionary<string, string> properties, IDictionary<string, double> measurements)
+        {
+            if (!_isInitialized)
+            {
+                return;
+            }
+
+            Dictionary<string, double> eventMeasurements = GetEventMeasures(measurements);
+            Dictionary<string, string> eventProperties = GetEventProperties(properties);
+
+            try
+            {
+                _client.TrackEvent(eventName, eventProperties, eventMeasurements);
+                _client.Flush();
+            }
+            catch (Exception) { }
+        }
+
+
+        private Dictionary<string, double> GetEventMeasures(IDictionary<string, double> measurements)
+        {
+            Dictionary<string, double> eventMeasurements = new Dictionary<string, double>(_commonMeasurements);
+            if (measurements != null)
+            {
+                foreach (var measurement in measurements)
+                {
+                    if (eventMeasurements.ContainsKey(measurement.Key))
+                    {
+                        eventMeasurements[measurement.Key] = measurement.Value;
+                    }
+                    else
+                    {
+                        eventMeasurements.Add(measurement.Key, measurement.Value);
+                    }
+                }
+            }
+            return eventMeasurements;
+        }
+
+        private Dictionary<string, string> GetEventProperties(IDictionary<string, string> properties)
+        {
+            Dictionary<string, string> eventProperties = new Dictionary<string, string>(_commonProperties);
```

Comment:
Is it necessary to declare the new dictionary in the case when no additional properties are specified?  It seems heavy handed to be declaring and coping the commonProperties in this case.  If I read the code correctly, this is the only code path that is exercised currently.